### DISCUSSION TITLE
Minor: Malformed JSON due \r in OtaUrl?

### DISCUSSION THREAD

**Discussion by fdcastel** [arendst/Tasmota#17233] [Q&A]
Flashed a new device with `tasmota32-zbbrdgpro.bin`. 

Default `OtaUrl` value contains a `\r` at end:

```
02:16:25.906 CMD: otaurl
02:16:25.911 RSL: RESULT = {"OtaUrl":"http://ota.tasmota.com/tasmota32/tasmota32-zbbrdgpro.bin\r"}
```

This makes the `status 0` command return an invalid JSON (my parser rejects it due un-escaped `0x0D` in string):

```json
{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/tasmota32-zbbrdgpro.bin
","RestartReason":"Software reset CPU","Uptime":"0T05:49:03","...":"..."}
```

Fix the default value or fix JSON serialization?

Workaround: Manually set `OtaUrl` via console with

```
Otaurl http://ota.tasmota.com/tasmota32/tasmota32-zbbrdgpro.bin
```

**Comment by barbudor** [comment]
So far I can't see anywhere in the code where would have happened
Are you 100% that it was there from the beginning?
Did you used any admin tool between the moment you flashed and the moment you discovered that?

**Reply by sfromis** [reply]
> un-escaped `0x0D` in string

This part of it looks like "some process" was using a line with cr/lf (`0x0A0D`) as line terminator (common Windows convention), and "some other process" was picking it up the lf as line terminator (common non-Windows convention), seeing the cr as part of the data.

Such issues might pop up if a Windows tool like Notepad was used to edit a file. Where and when is another question.....

**Reply by sfromis** [reply]
Anyway, I've confirmed the issue with a fresh flash of current dev `tasmota32-zbbrdgpro.factory.bin` without any setup or tools being run after the flashing:
```
00:02:11.553 CMD: otaurl
00:02:11.556 RSL: RESULT = {"OtaUrl":"http://ota.tasmota.com/tasmota32/tasmota32-zbbrdgpro.bin\r"}
```

**Reply by fdcastel** [reply]
My bad. Sorry for the lack of information:

- I bought a brand new Sonoff Zigbee Bridge PRO 
- Flashed it using [Tasmota Web Installer](https://tasmota.github.io/install/) 
- Flashed the coordinator firmware using [NotEnoughTech instructions](https://notenoughtech.com/home-automation/tasmota-on-sonoff-zb-bridge-pro/#flash) according the instructions from [Blackadder site](https://zigbee.blakadder.com/Sonoff_ZBBridge-P.html)
- All these operations were done from a Chrome Browser / Windows 10

The steps above reproduce the problem faithfully. I bought 2 devices. Flashed them exactly as above. No extra tool used. The problem occurred in both them.



> Are you 100% that it was there from the beginning?

Unfortunately, yes.



> Did you used any admin tool between the moment you flashed and the moment you discovered that?

Short answer: Nope. Just the web installer and the consoles (classic and berry).



I found the problem when I ran some custom scripts of mine. But they only QUERY the devices.

The first thing my script does is to send a `STATUS 0` to fetch all possible information from the device. At this point the .NET System.Text.Json parser crashed pointing the problem.



I have 2 devices. Neither is being used in production. I'm available to run any other tests, if you need.

**Reply by sfromis** [reply]
@Jason2866 Do you know where the extraneous cr is coming from?

**Reply by Jason2866** [reply]
@sfromis Good question, where is the OTA URL generated? It is not defined in `tasmota_configurations.h`
There is no change in this part afaik

**Reply by arendst** [reply]
I searched the code for ``otaurl`` and didn't find any clue either. Does it only happen on the zigbee bridge pro image or on others too?

**Reply by Jason2866** [reply]
it is from here https://github.com/tasmota/autoconf/blob/main/raw/esp32/Sonoff_ZBPro/init.bat
set from Autoconfig

**Reply by sfromis** [reply]
Indeed that file appears to be Windows-broken, with line delimiter cr/lf

**Reply by s-hadinger** [reply]
Indeed, I just fixed it.
Lesson learned, don't create autoconf files from Windows (can't happen to me :)

**Reply by s-hadinger** [reply]
I will add to my todo list to filter out any CR character as well

**Reply by Jason2866** [reply]
Dont understand how this CR/LF got in the file? I dont use Windows over a year. The change is younger.

**Reply by s-hadinger** [reply]
It must have been a copy/paste from another one. So I suspect that we have other files with CRLF.

**Reply by Jason2866** [reply]
Mhh, have to search/check if github has a possibility to auto correct that.

**Reply by fdcastel** [reply]
Thanks guys, you rock! ❤️ 

About the CR/LF problem: on my machine (Windows) the latest commit from Stephan shows **absolutely no differences** (for me):

![image](https://user-images.githubusercontent.com/629408/205107056-85c72549-0847-4e14-bdbf-a7c17c4dc1c4.png)

But that's probably due my [git configuration](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings) on Windows.

Anyway the build output should be the same no matter what operating system is used for commits.

_Just my 2 cents_: Maybe a review from these build scripts should be in order? As it is now it seems very fragile (and probably will bite us back again in the future).

Thank you for this amazing project! My entire house run on Tasmota.

**Comment by Jason2866** [comment]
This hopefully works https://github.com/tasmota/autoconf/blob/main/.gitattributes
